sql/core/src/main/scala/org/apache/spark/sql/execution/OptimizeMetadataOnly.scala

```diff
@@ -0,0 +1,133 @@
+import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, LogicalRelation}
+
+/**
+ * When scanning only partition columns, get results based on partition data without scanning files.
+ * It's used for operators that only need distinct values. Currently only [[Aggregate]] operator
+ * which satisfy the following conditions are supported:
+ * 1. aggregate expression is partition columns.
+ *  e.g. SELECT col FROM tbl GROUP BY col.
+ * 2. aggregate function on partition columns with DISTINCT.
+ *  e.g. SELECT count(DISTINCT col1) FROM tbl GROUP BY col2.
+ * 3. aggregate function on partition columns which have same result w or w/o DISTINCT keyword.
+ *  e.g. SELECT Max(col2) FROM tbl GROUP BY col1.
+ */
+case class OptimizeMetadataOnly(
+    catalog: SessionCatalog,
+    conf: CatalystConf) extends Rule[LogicalPlan] {
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!conf.optimizerMetadataOnly) {
+      return plan
+    }
+
+    plan.transform {
+      case a @ Aggregate(_, aggExprs, child @ PartitionedRelation(partAttrs, relation)) =>
+        if (a.references.subsetOf(partAttrs)) {
+          val aggFunctions = aggExprs.flatMap(_.collect {
+            case agg: AggregateExpression => agg
+          })
+          val isPartitionDataOnly = aggFunctions.isEmpty || aggFunctions.forall { agg =>
+            agg.isDistinct || (agg.aggregateFunction match {
+              case _: Max => true
+              case _: Min => true
+              case _ => false
+            })
+          }
+          if (isPartitionDataOnly) {
+            a.withNewChildren(Seq(usePartitionData(child, relation)))
+          } else {
+            a
+          }
+        } else {
+          a
+        }
+    }
+  }
+
+  private def usePartitionData(child: LogicalPlan, relation: LogicalPlan): LogicalPlan = {
+    child transform {
+      case plan if plan eq relation =>
+        relation match {
+          case l @ LogicalRelation(fsRelation: HadoopFsRelation, _, _) =>
+            val partColumns = fsRelation.partitionSchema.map(_.name.toLowerCase).toSet
+            val partAttrs = l.output.filter(a => partColumns.contains(a.name.toLowerCase))
+            val partitionData = fsRelation.location.listFiles(Nil)
+            LocalRelation(partAttrs, partitionData.map(_.values))
+
+          case relation: CatalogRelation =>
+            val partColumns = relation.catalogTable.partitionColumnNames.map(_.toLowerCase).toSet
+            val partAttrs = relation.output.filter(a => partColumns.contains(a.name.toLowerCase))
+            val partitionData = catalog.listPartitions(relation.catalogTable.identifier).map { p =>
+              InternalRow.fromSeq(partAttrs.map { attr =>
+                Cast(Literal(p.spec(attr.name)), attr.dataType).eval()
+              })
+            }
+            LocalRelation(partAttrs, partitionData)
+
+          case _ => throw new IllegalStateException()
+        }
+    }
+  }
+
+  object PartitionedRelation {
```

Comment:
we should also explain what patterns are acceptable, e.g. filter and project with deterministic expressions.